### ISSUE TITLE
Failed to communicate with API server: the server has received too many requests and has asked us to try again later (get componentstatuses.meta.k8s.io)

### ISSUE THREAD

**Issue by krishofmans** [rancher/rancher#16047]
**Rancher versions:**
rancher/rancher: 2.0.8 & 2.1.0
rancher/rancher-agent: rancher-agent:v2.1.0

**Infrastructure Stack versions:**
healthcheck: 
ipsec:
network-services:
scheduler:
kubernetes (if applicable): v1.11.3-rancher1-1 (we had it on previous versions, upgrading was our attempt to resolve these issues).

**Docker version: (`docker version`,`docker info` preferred)**
```Containers: 25
 Running: 19
 Paused: 0
 Stopped: 6
Images: 104
Server Version: 17.03.2-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 333
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 4ab9917febca54791c5f071a9d1f404867857fcc
runc version: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe
init version: N/A (expected: 949e6facb77383876aeff8a6944dde66b3089574)
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-36-generic
Operating System: Ubuntu 18.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.66 GiB
Name: k8s-development-3-1
ID: 4L3H:V32B:5BZ2:X6PJ:QDHP:3P4V:4CQR:EVYL:UL32:J5U4:HXJD:N5Y7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
```
root@k8s-development-3-1:/opt/k8s-storage# uname -r
4.15.0-36-generic
root@k8s-development-3-1:/opt/k8s-storage# cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
Hosts om vmware esxi 6.7

**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Single node 

**Environment Template: (Cattle/Kubernetes/Swarm/Mesos)**
Kubernetes

**Steps to Reproduce:**
We have a 5 node (3 all, 2 worker) kubernetes cluster running for a couple of months. It has quite a dynamic workload as it's mostly used to run builds using a statefulset of 2 gitlab runners (using anti affinity to not run on the same physical host). These constantly create and destroy pods triggered from the gitlab ci server.

I am not confident that this is the root cause. But the outages occur mostly during working hours.

There is other load, but mostly stateless java applications.

**Results:**
1. The api becomes unavailable
2. The node itself becomes unavailable
![image](https://user-images.githubusercontent.com/999774/46721722-02cc0d80-cc74-11e8-88fd-33b951f332cd.png)
3. Workloads are still running, but no longer created (due to the api being unavailable)

Disconnecting the node works as a workaround, but we've replaced the node by another vm and it becomes unstable again.

**Workaround:**
When we remove node 3, the cluster becomes available again. We've never seen this issue manifest itself with a 4 node cluster with 2 worker and 2 all nodes. But we know that we risk the etcd split brain issue at that time.

I've attached the kubelet log from node 3, but I can supply any log from any container required.
Issue starts at line 2320 where the error below starts appearing.
```
error getting node \“k8s-development-3-1\“: the server has received too many requests and has asked us to try again later
```
[kubelet.log](https://github.com/rancher/rancher/files/2463901/kubelet.log)

**Comment by superseb** [comment]
Do heavy workloads run on the nodes running the kube-apiserver? Does it also happen when you separate them? Are pods limited in resources or not?

**Comment by krishofmans** [comment]
De heavy workloads are distributed over all nodes. The gitlab runner pods are limited in resources: 
![image](https://user-images.githubusercontent.com/999774/46748475-ef409700-ccb3-11e8-8f63-0ef176a742ed.png)

We have not yet tried separating them. Is this a best practice to keep the api nodes available only as an api node?

Resource pressure does not seem to impact this problem too much as it happened under a load with lesser builds as well.

Now that we have removed the third node it has been stable again whole day without issues, also running builds. And a similar workload had run fine for 2 months before problems started showing, I think it was around ~2.0.8 upgrade, but we can't verify this for sure.

**Comment by krishofmans** [comment]
Today the cluster failed again with 4 nodes. Investigation showed that filebeat was throwing errors and hitting it's 200mb memory limit. But no apparent cause. It is a suspect though, since it will use the kube-api to add metadata.

Also the node had a lot of left over containers from builds that used docker-in-docker. We've cleaned up those and are adding --rm to any dind operation.

Then we restructured the cluster to have 3 separate etcd/api nodes and are now assessing the situation to see if the problem comes back.

![image](https://user-images.githubusercontent.com/999774/46944384-2c20db00-d072-11e8-8979-b268d19f764e.png)

**Comment by krishofmans** [comment]
I believe we have tracked down the root cause, being this issue in RKE: https://github.com/rancher/rke/issues/866#issuecomment-420019920

By already splitting the nodes the certificates were correct, but we did not delete the kube metrics pod during this process. We could see the logs being flooded by:
```
[x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "kube-apiserver-requestheader-ca"), x509: certificate signed by unknown authority]
```

While now it looks much saner like this:
![image](https://user-images.githubusercontent.com/999774/46951879-84170c00-d089-11e8-949b-6f9f593e5f69.png)

My assumption: the kube metrics server was spamming the api with an incorrect certificate, leading to rate throttling.
I'm giving this 72hours, we usually see failure much sooner than that, after that I will close this ticket.

**Comment by krishofmans** [comment]
Unfortunately we've had an unresponsive node again this morning.

Restarting docker on the management node made the cluster available again.

We allowed a prometheus to start scraping kubernetes metrics again, but we're not sure if this is the cause of the api becoming unavailable.

![image](https://user-images.githubusercontent.com/999774/47207731-44ec0200-d38c-11e8-9dd0-fbcb7fea1f9c.png)

it scrapes some more cadvisors and then some pods with prometheus annotations, but I don't think they should generate a lot of kube api calls.

Some nightly builds terminated with this message:
```
Error from server (InternalError): an error on the server ("error dialing backend: x509: certificate is valid for 192.168.26.139, 192.168.26.10, 192.168.26.246, 127.0.0.1, 10.43.0.1, not 192.168.26.230") has prevented the request from succeeding
```

the ip where it was not valid is the node that was unresponsive. So perhaps the certification rotation hasn't completely recovered yet?

**Comment by krishofmans** [comment]
Closing this ticket, as it seems for this the cause was: https://github.com/rancher/rke/issues/866#issuecomment-420019920

Other issues seem to have been caused by a docker bug (https://github.com/moby/moby/issues/35091)

This seems to have taken away all the issues we have currently, we did see some stacktraces related to possibly this: https://github.com/kubernetes-incubator/metrics-server/issues/132 and assume the fix will come when 1.12 is considered stable.

Customer is also considering commercial rancher support at this time.
